### PR TITLE
Rewrite get_default_modules()

### PR DESCRIPTION
**Pull request by sgallagher**
The previous version was needlessly complicated and difficult to
follow. This patch replaces it with a lookup of the latest
versions of each of the default streams and a recursive lookup of
their dependencies.

Also adds a `--debug` flag and some debugging messages.

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

### CODE REVIEW

File: splitter.py [hunk 1]
```diff
@@ -184,6 +185,70 @@ def validate_filenames(directory, repoinfo):
...
+    # At least one of the dependency array entries must exist in the repo
+    found_dep = False
+    for dep in deps:
+        # Within an array entry, all of the modules must be present in the index
+        found_all_modules = True
+        for modname in dep.get_runtime_modules():
+            # Ignore "platform" because it's special
+            if modname == "platform":
+                logging.debug('Skipping platform')
+                continue
```
Comment (Conan-Kudo) [inline splitter.py:213]: Is `platform` the only case we want to ignore?

Comment (sgallagher) [reply splitter.py:213]: `platform` is the only special-case for runtime. If that ever changes, we'll need to modify this (but we'd also need to modify it in many other projects/libraries)

**Comment by Conan-Kudo** [comment]
@sgallagher I might be reading this a bit naively, but it seems like this code assumes that the modules and module dependencies are resolvable within a repository. Is that a correct assumption?

**Comment by sgallagher** [comment]
> @sgallagher I might be reading this a bit naively, but it seems like this code assumes that the modules and module dependencies are resolvable within a repository. Is that a correct assumption?

It is working under the assumption that we are interacting with the RHEL 8 AppStream repository, which contains all modules shipped by RHEL 8.

If we want to make a different assumption, then I guess I'd rework the patch to just ensure we get whatever subset of the streams are available in this repository.

**Comment by Conan-Kudo** [comment]
@sgallagher I think for now this is fine, but I'm more concerned about dealing with layered repositories that contain modules that describe module dependencies in base layer repositories in the future.

**Review by Conan-Kudo — APPROVED** [review]

**Comment by sgallagher** [comment]
@Conan-Kudo Yeah, if we want this to be reusable for general purposes (such as working with Fedora updates repos), we should probably make the change I mentioned above. But if we only care about RHEL 8, this is probably fine.

@smooge Can I get your opinion on this matter?

**Comment by sgallagher** [comment]
I pushed two minor changes (amended). The first was to fix up a couple formatting issues and the second changes how `_get_recursive_dependencies()` saves its results back to the `all_deps` variable. Previously, if we ended up skipping a default stream because its dependencies didn't resolve, we were still including whatever subset was available. Now we exclude that, so if we skip a default stream we don't carry it along.

The previous behavior was safe, just potentially results in carrying more content than is desired.

**Comment by sgallagher** [comment]
And one more change, this one in its own patch. It adds a `--ignore-missing-default-deps` argument that will allow us to retrieve as much as possible for the default streams, even if some of their dependencies are coming from another repository.

@Conan-Kudo Would you mind re-reviewing, please?

**Review by Conan-Kudo — APPROVED** [review]
At a glance, things look good to me. I don't have a sample on hand to validate this new logic, but at least it doesn't look broken.

**Pull request merged**